### PR TITLE
268 fix unsaved form warning

### PR DESCRIPTION
**Pull request by danieldunderfelt**
Issue: https://github.com/HSLdevcom/bultti/issues/268

No server PR.

Refactor the unsaved form feature and switch to react-router to make it more integrated. Reach-router cannot prevent transitions.

- Refactor unsaved form feature
- Switch to react-router
- Implement navigation blockers with react-router

### CODE REVIEW

**Review by Divinare — CHANGES_REQUESTED** [review]
Nice work! I like how the prompt is implemented. Left just a few minor comments

Testing:

* Prompting seems to work in each form!

* New idea (this should be implemented in a new card): how about adding "disabledText" props to a button so that user could hover over grey button and see the text OR click the button and get notification
  * e.g. [greyed lähetä tarkastettavaksi] --> "Sinulla on tallentamattomia muutoksia. Tallenna tai hylkää ne ennen lähetystä tarkastettavaksi"

File: src/util/urlValue.ts [hunk 1]
```diff
@@ -126,22 +77,37 @@ function excludeQueryParams(queryString = history.location.search) {
   return query
 }
 
-export const pathWithQuery = (path = '', location?: Location | string) => {
+interface HasQueryString {
+  search: string
+}
+
+// Provide a Location object with a `search` param or the full URL with the query string.
+export function pathWithQuery(path = '', location?: HasQueryString | string) {
```
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/util/urlValue.ts:85]: Excluded query params really doesn't need to come from env. And I don't see the value in combining the two functions.

File: src/util/useAuth.ts [hunk 2]
```diff
@@ -102,9 +106,9 @@ export const useAuth = (): [AuthState, boolean] => {
 
     if (codeUrlParam && authState === AuthState.UNAUTHENTICATED) {
       setAuthState(AuthState.PENDING)
-      setUrlValue('code', null)
-      setUrlValue('isTest', null)
-      setUrlValue('role', null)
+      setUrlValue(history, 'code', null)
+      setUrlValue(history, 'isTest', null)
+      setUrlValue(history, 'role', null)
```
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/util/useAuth.ts:111]: Sure, but this isn't really meant to be a general interface for the search string. It's for storing state values in the URL, and the mechanic happens to be the search string. I don't feel that we need to follow any official naming, and it's better to keep it as a "domain" name to not give the impression that it IS a general interface for the search string.

File: src/state/stateContext.tsx [hunk 3]
```diff
@@ -1,4 +1,36 @@
...
+
+export const StateProvider: React.FC = ({ children }) => {
+  const history = useHistory()
+  const state = useMemo(() => createState(history, initializers), [])
+
+  const onAuthError = useCallback(() => {
+    removeAuthToken()
+    state.actions.user(null)
+    state.actions.notifications.add({
+      message: 'Authentication expired or invalid. Please log in again.',
```
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/state/stateContext.tsx:24]: Yes, I'll add it!

File: src/util/urlValue.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/util/urlValue.ts:85]: Changed the name of the method to `getPathWithSearch` to reflect that it works with the search string.

**Pull request merged**